File: src/Z.Reflection/GetDeclaration/MemberInfo.GetDeclaration.cs
// Description: C# Extension Methods | Enhance the .NET Framework and .NET Core with over 1000 extension methods.
// Website & Documentation: https://csharp-extension.com/
// Issues: https://github.com/zzzprojects/Z.ExtensionMethods/issues
// License (MIT): https://github.com/zzzprojects/Z.ExtensionMethods/blob/master/LICENSE
// More projects: https://zzzprojects.com/
// Copyright © ZZZ Projects Inc. All rights reserved.
using System;
using System.Reflection;

public static partial class Extensions
{
    /// <summary>A MemberInfo extension method that gets a declaraction.</summary>
    /// <param name="this">The @this to act on.</param>
    /// <returns>The declaraction.</returns>
    public static string GetDeclaraction(this MemberInfo @this)
    {
        switch (@this.MemberType)
        {
            case MemberTypes.Field:
                return ((FieldInfo) @this).GetDeclaraction();
            case MemberTypes.Property:
                return ((PropertyInfo) @this).GetDeclaraction();
            case MemberTypes.Constructor:
                return ((ConstructorInfo) @this).GetDeclaraction();
            case MemberTypes.Method:
                return ((MethodInfo) @this).GetDeclaraction();
            case MemberTypes.TypeInfo:
                return ((Type) @this).GetDeclaraction();
            case MemberTypes.NestedType:
                return ((Type) @this).GetDeclaraction();
            case MemberTypes.Event:
                return ((EventInfo) @this).GetDeclaraction();
            default:
                return null;
        }
    }
}

File: src (VB.NET)/Z.ExtensionMethods.VB/Z.Reflection/GetDeclaration/MemberInfo.GetDeclaration.vb

' Copyright (c) 2015 ZZZ Projects. All rights reserved
' Licensed under MIT License (MIT) (https://github.com/zzzprojects/Z.ExtensionMethods)
' Website: http://www.zzzprojects.com/
' Feedback / Feature Requests / Issues : http://zzzprojects.uservoice.com/forums/283927
' All ZZZ Projects products: Entity Framework Extensions / Bulk Operations / Extension Methods /Icon Library

Imports System.Reflection

Public Module Extensions_808

	<System.Runtime.CompilerServices.Extension> _
	Public Function GetDeclaraction(this As MemberInfo) As String
		Select Case this.MemberType
			Case MemberTypes.Field
				Return DirectCast(this, FieldInfo).GetDeclaraction()
			Case MemberTypes.[Property]
				Return DirectCast(this, PropertyInfo).GetDeclaraction()
			Case MemberTypes.Constructor
				Return DirectCast(this, ConstructorInfo).GetDeclaraction()
			Case MemberTypes.Method
				Return DirectCast(this, MethodInfo).GetDeclaraction()
			Case MemberTypes.TypeInfo
				Return DirectCast(this, Type).GetDeclaraction()
			Case MemberTypes.NestedType
				Return DirectCast(this, Type).GetDeclaraction()
			Case MemberTypes.[Event]
				Return DirectCast(this, EventInfo).GetDeclaraction()
			Case Else
				Return Nothing
		End Select
	End Function
End Module


